File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.maxH = 0
        def height(root):
            if root is None: return 0
        
            lh = height(root.left)        
            rh = height(root.right)

            self.maxH = max(self.maxH, lh+rh)

            return 1 + max(lh,rh)
        height(root)
        return self.maxH
    

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    let max = 0
    
    const rec = (node) => {
        if (node==null){
            return 0;
        }
        
        let lh = rec(node.left, max);
        let rh = rec(node.right, max);
        
        max = Math.max(max, lh+rh)
        
        return 1+Math.max(lh, rh)
    }
    
    rec(root)
    return max
};